diff --git a/core/assets/scaffold/files/default.settings.php b/core/assets/scaffold/files/default.settings.php
index d0fbcd3512..3a9025ed1f 100644
--- a/core/assets/scaffold/files/default.settings.php
+++ b/core/assets/scaffold/files/default.settings.php
@@ -578,10 +578,15 @@
  * A custom theme for the offline page:
  *
  * This applies when the site is explicitly set to maintenance mode through the
- * administration page or when the database is inactive due to an error.
- * The template file should also be copied into the theme. It is located inside
+ * administration page.
+ * The template file should also be copied into the theme. It is located at
  * 'core/modules/system/templates/maintenance-page.html.twig'.
  *
+ * This applies also when the database is inactive due to an error
+ * or when a fatal error is thrown.
+ * The template file should also be copied into the theme. It is located at
+ * 'core/modules/system/templates/maintenance-page--offline.html.twig'.
+ *
  * Note: This setting does not apply to installation and update pages.
  */
 # $settings['maintenance_theme'] = 'bartik';
diff --git a/core/includes/errors.inc b/core/includes/errors.inc
index 045b2dd417..5c381873d4 100644
--- a/core/includes/errors.inc
+++ b/core/includes/errors.inc
@@ -260,7 +260,7 @@ function _drupal_log_error($error, $fatal = FALSE) {
       // We fallback to a maintenance page at this point, because the page generation
       // itself can generate errors.
       // Should not translate the string to avoid errors producing more errors.
-      $message = 'The website encountered an unexpected error. Please try again later.' . '<br />' . $message;
+      $message = 'The website encountered an unexpected error. Please try again later.' . '<br /><br />' . $message;
 
       if ($is_installer) {
         // install_display_output() prints the output and ends script execution.
@@ -272,6 +272,16 @@ function _drupal_log_error($error, $fatal = FALSE) {
         exit;
       }
 
+      $html = Error::renderFatalError([
+        'title' => 'Service unavailable',
+        'content' => $message,
+        'displayable' => error_displayable($error),
+      ]);
+      if (!empty($html)) {
+        $message = $html;
+        $response->headers->set('Content-Type', 'text/html');
+      }
+
       $response->setContent($message);
       $response->setStatusCode(500, '500 Service unavailable (with message)');
 
diff --git a/core/lib/Drupal/Core/EventSubscriber/FinalExceptionSubscriber.php b/core/lib/Drupal/Core/EventSubscriber/FinalExceptionSubscriber.php
index ef0123e08c..712a5dc55e 100644
--- a/core/lib/Drupal/Core/EventSubscriber/FinalExceptionSubscriber.php
+++ b/core/lib/Drupal/Core/EventSubscriber/FinalExceptionSubscriber.php
@@ -27,9 +27,10 @@
  * handled by a single exception subscriber:: serialization.exception.default.
  *
  * This exception subscriber runs after all the above (it has a lower priority),
- * which makes it the last-chance exception handler. It always sends a plain
- * text response. If it's a displayable error and the error level is configured
- * to be verbose, then a helpful backtrace is also printed.
+ * which makes it the last-chance exception handler. It sends a html response
+ * if a request format is html and a plain text otherwise. If it's a displayable
+ * error and the error level is configured to be verbose, then a helpful
+ * backtrace is also printed.
  */
 class FinalExceptionSubscriber implements EventSubscriberInterface {
   use StringTranslationTrait;
@@ -125,6 +126,18 @@ public function onException(ExceptionEvent $event) {
     $content_type = $event->getRequest()->getRequestFormat() == 'html' ? 'text/html' : 'text/plain';
     $content = $this->t('The website encountered an unexpected error. Please try again later.');
     $content .= $message ? '</br></br>' . $message : '';
+
+    if ($content_type == 'text/html') {
+      $html = Error::renderFatalError([
+        'title' => $this->t('Service unavailable'),
+        'content' => $content,
+        'displayable' => $this->isErrorDisplayable($error),
+      ]);
+      if (!empty($html)) {
+        $content = $html;
+      }
+    }
+
     $response = new Response($content, 500, ['Content-Type' => $content_type]);
 
     if ($exception instanceof HttpExceptionInterface) {
diff --git a/core/lib/Drupal/Core/Utility/Error.php b/core/lib/Drupal/Core/Utility/Error.php
index 8cec41f8ba..924b03da96 100644
--- a/core/lib/Drupal/Core/Utility/Error.php
+++ b/core/lib/Drupal/Core/Utility/Error.php
@@ -2,11 +2,16 @@
 
 namespace Drupal\Core\Utility;
 
-use Drupal\Component\Render\FormattableMarkup;
+use Drupal\Core\DependencyInjection\ContainerNotInitializedException;
+use Drupal\Core\Extension\ExtensionDiscovery;
+use Drupal\Core\Site\Settings;
 use Drupal\Component\Utility\Xss;
+use Drupal\Component\Render\FormattableMarkup;
 use Drupal\Core\Database\Database;
 use Drupal\Core\Database\DatabaseExceptionWrapper;
 use Drupal\Core\Database\Log;
+use Twig\Environment;
+use Twig\Loader\ArrayLoader;
 
 /**
  * Drupal error utility class.
@@ -95,6 +100,69 @@ public static function renderExceptionSafe($exception) {
     return new FormattableMarkup('%type: @message in %function (line %line of %file). <pre class="backtrace">@backtrace</pre>', $decode);
   }
 
+  /**
+   * Renders fatal error page from twig template using Symfony twig engine.
+   *
+   * @param array $context
+   *   Variables to be passed to twig template.
+   *
+   * @return string
+   *   An html code with rendered fatal error page.
+   */
+  public static function renderFatalError(array $context) {
+    $template_path = '/templates/maintenance-page--offline.html.twig';
+    $system_path = 'core/modules/system';
+    $theme = '';
+
+    // Get offline theme from settings.php and check if the template exists.
+    try {
+      $theme = Settings::get('maintenance_theme', '');
+      if (!$theme) {
+        $theme_path = $system_path;
+      }
+      else {
+        $theme_path = \Drupal::service('extension.list.theme')->getPath($theme);
+      }
+    }
+    catch (ContainerNotInitializedException $e) {
+      // The maintenance theme is set but the container doesn't exist
+      // since the database is inactive. Hence there are no services available
+      // to retrieve a maintenance theme path. The path can be obtained by using
+      // ExtensionDiscovery::scan() but the app root should be guessed first
+      // in the same way as DrupalKernel::guessApplicationRoot() does.
+      $app_root = dirname(substr(__DIR__, 0, -strlen(__NAMESPACE__)), 2);
+      $listing = new ExtensionDiscovery($app_root, FALSE, NULL, NULL);
+      // An empty profile directory prevents ExtensionDiscovery::scan()
+      // from calling \Drupal::installProfile() that needs working container.
+      $listing->setProfileDirectories([]);
+      $themes = $listing->scan('theme');
+      $theme_path = isset($themes[$theme]) ? $themes[$theme]->getPath() : $system_path;
+      if ($context['displayable']) {
+        $context['content'] = $context['content'] . "<pre>" . $e . "</pre>";
+      }
+    }
+    catch (\Throwable $error) {
+      // Handle any other cases.
+      $theme_path = $system_path;
+      if ($context['displayable']) {
+        $context['content'] = $context['content'] . "<pre>" . $error . "</pre>";
+      }
+    }
+
+    $path = $theme_path . $template_path;
+    if (!file_exists($path)) {
+      $path = $system_path . $template_path;
+    }
+
+    // Directly use Symfony twig engine without Drupal wrapper to minimize
+    // possibility of nested exception.
+    $template = file_get_contents($path);
+    $loader = new ArrayLoader(['maintenance_page_offline' => $template]);
+    $environment = new Environment($loader);
+
+    return $environment->render('maintenance_page_offline', $context);
+  }
+
   /**
    * Gets the last caller from a backtrace.
    *
diff --git a/core/modules/big_pipe/tests/src/Functional/BigPipeTest.php b/core/modules/big_pipe/tests/src/Functional/BigPipeTest.php
index ab7b26c0a8..624dacaec8 100644
--- a/core/modules/big_pipe/tests/src/Functional/BigPipeTest.php
+++ b/core/modules/big_pipe/tests/src/Functional/BigPipeTest.php
@@ -213,9 +213,8 @@ public function testBigPipe() {
     // The 'edge_case__html_exception' case throws an exception.
     $this->assertRaw('The website encountered an unexpected error. Please try again later');
     $this->assertRaw('You are not allowed to say llamas are not cool!');
-    // Check that stop signal and closing body tag are absent.
+    // Check that stop signal is absent.
     $this->assertNoRaw(BigPipe::STOP_SIGNAL);
-    $this->assertNoRaw('</body>');
     // The exception is expected. Do not interpret it as a test failure.
     unlink($this->root . '/' . $this->siteDirectory . '/error.log');
   }
@@ -285,7 +284,6 @@ public function testBigPipeNoJs() {
     // The 'edge_case__html_exception' case throws an exception.
     $this->assertRaw('The website encountered an unexpected error. Please try again later');
     $this->assertRaw('You are not allowed to say llamas are not cool!');
-    $this->assertNoRaw('</body>');
     // The exception is expected. Do not interpret it as a test failure.
     unlink($this->root . '/' . $this->siteDirectory . '/error.log');
   }
diff --git a/core/modules/system/templates/maintenance-page--offline.html.twig b/core/modules/system/templates/maintenance-page--offline.html.twig
new file mode 100644
index 0000000000..d732bf7b89
--- /dev/null
+++ b/core/modules/system/templates/maintenance-page--offline.html.twig
@@ -0,0 +1,41 @@
+{#
+/**
+ * @file
+ * Default theme implementation to display a single Drupal page while offline.
+ *
+ * @see template_preprocess_maintenance_page()
+ *
+ * @ingroup themeable
+ */
+#}
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>
+    {% if title %}
+      {{ title }}
+    {% else %}
+      Service Unavailable
+    {% endif %}
+  </title>
+</head>
+<body class="maintenance-page layout-no-sidebars path-frontpage">
+  <div id="page-wrapper">
+    <div id="page">
+      <div id="main-wrapper">
+        <div id="main" class="clearfix">
+          <main id="content" class="column" role="main">
+            <section class="section">
+              <a id="main-content"></a>
+              {% if title %}
+                <h1 class="title" id="page-title">{{ title }}</h1>
+              {% endif %}
+              {{ content|striptags('<p>,<br>,<em>,<pre>')|raw }}
+            </section>
+          </main>
+        </div>
+      </div>
+    </div>
+  </div>
+</body>
+</html>
diff --git a/core/modules/system/tests/src/Functional/System/MaintenancePageOfflineTest.php b/core/modules/system/tests/src/Functional/System/MaintenancePageOfflineTest.php
new file mode 100644
index 0000000000..2b0de0be9c
--- /dev/null
+++ b/core/modules/system/tests/src/Functional/System/MaintenancePageOfflineTest.php
@@ -0,0 +1,210 @@
+<?php
+
+namespace Drupal\Tests\system\Functional\System;
+
+use Drupal\Core\Database\Database;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Tests if the Maintenance page is served when the site is offline.
+ *
+ * @group system
+ */
+class MaintenancePageOfflineTest extends BrowserTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected $defaultTheme = 'test_theme';
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    $settings_filename = $this->siteDirectory . '/settings.php';
+    chmod($settings_filename, 0777);
+    $settings_php = file_get_contents($settings_filename);
+    // Ensure we can test errors rather than being caught in
+    // \Drupal\Core\Test\HttpClientMiddleware\TestHttpClientMiddleware.
+    $settings_php .= "\ndefine('SIMPLETEST_COLLECT_ERRORS', FALSE);\n";
+    file_put_contents($settings_filename, $settings_php);
+  }
+
+  /**
+   * Prepare settings values before a test case.
+   *
+   * @param string $maintenance_theme
+   *   The name of a maintenance theme. Empty if there is no maintenance theme.
+   * @param string $error_level
+   *   The name of an error level.
+   * @param bool $active_database
+   *   TRUE if a database should be active.
+   * @param bool $valid_hash_salt
+   *   TRUE if a hash_salt should be valid.
+   */
+  protected function prepareCaseSettings($maintenance_theme, $error_level, $active_database = TRUE, $valid_hash_salt = TRUE) {
+    $settings = [];
+    if (!empty($maintenance_theme)) {
+      $settings['settings']['maintenance_theme'] = (object) [
+        'value' => $maintenance_theme,
+        'required' => TRUE,
+      ];
+    }
+    $settings['config']['system.logging']['error_level'] = (object) [
+      'value' => $error_level,
+      'required' => TRUE,
+    ];
+    if (!$active_database) {
+      // Make a database inactive by setting an invalid password.
+      $connection_info = Database::getConnectionInfo();
+      $settings['databases']['default']['default']['password'] = (object) [
+        'value' => $connection_info['default']['password'] . $this->randomMachineName(),
+        'required' => TRUE,
+      ];
+    }
+    if (!$valid_hash_salt) {
+      // Set a hash_salt to invalid value.
+      $settings['settings']['hash_salt'] = (object) [
+        'value' => NULL,
+        'required' => TRUE,
+      ];
+    }
+    $this->writeSettings($settings);
+  }
+
+  /**
+   * Tests cases when settings.php contains invalid database settings.
+   *
+   * Tests if the maintenance offline page is served when settings.php
+   * contains invalid database settings.
+   */
+  public function testInvalidDatabaseSettings() {
+    // Open a frontpage without any error.
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(200);
+    $this->assertSession()->pageTextContains('Log in');
+
+    // CASE 1 - a maintenance theme's template is picked up
+    // and no errors are displayed.
+    $this->prepareCaseSettings('test_theme', ERROR_REPORTING_HIDE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A maintenance theme's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title test-theme"');
+    // A fatal error message and a backtrace should be hidden.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextNotContains('Access denied for user');
+    $this->assertSession()->responseNotContains('<pre class="backtrace">');
+
+    // CASE 2 - a maintenance theme's template is picked up
+    // and all errors with a backtrace are displayed.
+    $this->prepareCaseSettings('test_theme', ERROR_REPORTING_DISPLAY_VERBOSE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A maintenance theme's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title test-theme"');
+    // A fatal error message and a backtrace should be shown.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextContains('Access denied for user');
+    $this->assertSession()->responseContains('<pre class="backtrace">');
+
+    // CASE 3 - a system's template is picked up
+    // since a maintenance theme doesn't have the template
+    // and no errors are displayed.
+    $this->prepareCaseSettings('test_subtheme', ERROR_REPORTING_HIDE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A system's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title"');
+    // A fatal error message and a backtrace should be hidden.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextNotContains('Access denied for user');
+    $this->assertSession()->responseNotContains('<pre class="backtrace">');
+
+    // CASE 4 - a system's template is picked up
+    // since a maintenance theme is not set
+    // and all errors with a backtrace are displayed.
+    $this->prepareCaseSettings('', ERROR_REPORTING_DISPLAY_VERBOSE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A system's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title"');
+    // A fatal error message and a backtrace should be shown.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextContains('Access denied for user');
+    $this->assertSession()->responseContains('<pre class="backtrace">');
+  }
+
+  /**
+   * Tests cases when settings.php doesn't have a hash_salt defined.
+   *
+   * Tests if the maintenance offline page is served when settings.php
+   * originally did have a hash_salt defined, which is emptied out afterwards.
+   */
+  public function testRemovedHashSalt() {
+    // Open a frontpage without any error.
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(200);
+    $this->assertSession()->pageTextContains('Log in');
+
+    // CASE 1 - a maintenance theme's template is picked up
+    // and no errors are displayed.
+    $this->prepareCaseSettings('test_theme', ERROR_REPORTING_HIDE, TRUE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A maintenance theme's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title test-theme"');
+    // A fatal error message and a backtrace should be hidden.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextNotContains('Missing $settings[\'hash_salt\'] in settings.php');
+    $this->assertSession()->responseNotContains('<pre class="backtrace">');
+
+    // CASE 2 - a maintenance theme's template is picked up
+    // and all errors with a backtrace are displayed.
+    $this->prepareCaseSettings('test_theme', ERROR_REPORTING_DISPLAY_VERBOSE, TRUE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A maintenance theme's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title test-theme"');
+    // A fatal error message and a backtrace should be shown.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextContains('Missing $settings[\'hash_salt\'] in settings.php');
+    $this->assertSession()->responseContains('<pre class="backtrace">');
+
+    // CASE 3 - a system's template is picked up
+    // since a maintenance theme doesn't have the template
+    // and no errors are displayed.
+    $this->prepareCaseSettings('test_subtheme', ERROR_REPORTING_HIDE, TRUE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A system's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title"');
+    // A fatal error message and a backtrace should be hidden.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextNotContains('Missing $settings[\'hash_salt\'] in settings.php');
+    $this->assertSession()->responseNotContains('<pre class="backtrace">');
+
+    // CASE 4 - a system's template is picked up
+    // since a maintenance theme is not set
+    // and all errors with a backtrace are displayed.
+    $this->prepareCaseSettings('', ERROR_REPORTING_DISPLAY_VERBOSE, TRUE, FALSE);
+    $this->drupalGet('');
+    $this->assertSession()->statusCodeEquals(500);
+    // A system's offline template should be picked up.
+    $this->assertSession()->pageTextContains('Service unavailable');
+    $this->assertSession()->responseContains('<h1 class="title"');
+    // A fatal error message and a backtrace should be shown.
+    $this->assertSession()->pageTextContains('The website encountered an unexpected error. Please try again later.');
+    $this->assertSession()->pageTextContains('Missing $settings[\'hash_salt\'] in settings.php');
+    $this->assertSession()->responseContains('<pre class="backtrace">');
+  }
+
+}
diff --git a/core/modules/system/tests/themes/test_theme/templates/maintenance-page--offline.html.twig b/core/modules/system/tests/themes/test_theme/templates/maintenance-page--offline.html.twig
new file mode 100644
index 0000000000..5269ffb4b0
--- /dev/null
+++ b/core/modules/system/tests/themes/test_theme/templates/maintenance-page--offline.html.twig
@@ -0,0 +1,41 @@
+{#
+/**
+ * @file
+ * Default theme implementation to display a single Drupal page while offline.
+ *
+ * @see template_preprocess_maintenance_page()
+ *
+ * @ingroup themeable
+ */
+#}
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <title>
+    {% if title %}
+      {{ title }}
+    {% else %}
+      Service Unavailable
+    {% endif %}
+  </title>
+</head>
+<body class="maintenance-page layout-no-sidebars path-frontpage">
+  <div id="page-wrapper">
+    <div id="page">
+      <div id="main-wrapper">
+        <div id="main" class="clearfix">
+          <main id="content" class="column" role="main">
+            <section class="section">
+              <a id="main-content"></a>
+              {% if title %}
+                <h1 class="title test-theme" id="page-title">{{ title }}</h1>
+              {% endif %}
+              {{ content|striptags('<p>,<br>,<em>,<pre>')|raw }}
+            </section>
+          </main>
+        </div>
+      </div>
+    </div>
+  </div>
+</body>
+</html>
diff --git a/sites/default/default.settings.php b/sites/default/default.settings.php
index d0fbcd3512..3a9025ed1f 100644
--- a/sites/default/default.settings.php
+++ b/sites/default/default.settings.php
@@ -578,10 +578,15 @@
  * A custom theme for the offline page:
  *
  * This applies when the site is explicitly set to maintenance mode through the
- * administration page or when the database is inactive due to an error.
- * The template file should also be copied into the theme. It is located inside
+ * administration page.
+ * The template file should also be copied into the theme. It is located at
  * 'core/modules/system/templates/maintenance-page.html.twig'.
  *
+ * This applies also when the database is inactive due to an error
+ * or when a fatal error is thrown.
+ * The template file should also be copied into the theme. It is located at
+ * 'core/modules/system/templates/maintenance-page--offline.html.twig'.
+ *
  * Note: This setting does not apply to installation and update pages.
  */
 # $settings['maintenance_theme'] = 'bartik';
