diff --git a/core/modules/book/tests/src/FunctionalJavascript/BookJavascriptTest.php b/core/modules/book/tests/src/FunctionalJavascript/BookJavascriptTest.php
index 1c8616dcf1..717c890605 100644
--- a/core/modules/book/tests/src/FunctionalJavascript/BookJavascriptTest.php
+++ b/core/modules/book/tests/src/FunctionalJavascript/BookJavascriptTest.php
@@ -2,7 +2,6 @@
 
 namespace Drupal\Tests\book\FunctionalJavascript;
 
-use Behat\Mink\Exception\ExpectationException;
 use Drupal\Component\Render\FormattableMarkup;
 use Drupal\FunctionalJavascriptTests\WebDriverTestBase;
 use Drupal\node\Entity\Node;
@@ -64,7 +63,7 @@ public function testBookOrdering() {
     $this->assertGreaterThan($weight_select1->getValue(), $weight_select2->getValue());
 
     // Check that '1st page' precedes the '2nd page'.
-    $this->assertOrderInPage(['1st page', '2nd page']);
+    $this->assertSession()->responseContentHasOrder(['1st page', '2nd page']);
 
     // Check that the 'unsaved changes' text is not present in the message area.
     $this->assertSession()->pageTextNotContains('You have unsaved changes.');
@@ -84,7 +83,7 @@ public function testBookOrdering() {
     $this->assertSession()->pageTextContains('You have unsaved changes.');
 
     // Check that '2nd page' page precedes the '1st page'.
-    $this->assertOrderInPage(['2nd page', '1st page']);
+    $this->assertSession()->responseContentHasOrder(['2nd page', '1st page']);
 
     $this->submitForm([], 'Save book pages');
     $this->assertSession()->pageTextContains(new FormattableMarkup('Updated book @book.', ['@book' => $book->getTitle()]));
@@ -95,7 +94,7 @@ public function testBookOrdering() {
     $this->assertGreaterThan($page2->book['weight'], $page1->book['weight']);
 
     // Check again that '2nd page' is on top after form submit in the UI.
-    $this->assertOrderInPage(['2nd page', '1st page']);
+    $this->assertSession()->responseContentHasOrder(['2nd page', '1st page']);
 
     // Toggle row weight selects as visible.
     $page->findButton('Show row weights')->click();
@@ -125,7 +124,7 @@ public function testBookOrdering() {
     $this->assertSession()->pageTextContains(new FormattableMarkup('Updated book @book.', ['@book' => $book->getTitle()]));
 
     // Check that the '1st page' is first again.
-    $this->assertOrderInPage(['1st page', '2nd page']);
+    $this->assertSession()->responseContentHasOrder(['1st page', '2nd page']);
 
     // Check that page reordering was done in the backend for manual weight
     // field usage.
@@ -134,34 +133,6 @@ public function testBookOrdering() {
     $this->assertGreaterThan($page2->book['weight'], $page1->book['weight']);
   }
 
-  /**
-   * Asserts that several pieces of markup are in a given order in the page.
-   *
-   * @param string[] $items
-   *   An ordered list of strings.
-   *
-   * @throws \Behat\Mink\Exception\ExpectationException
-   *   When any of the given string is not found.
-   *
-   * @todo Remove this once https://www.drupal.org/node/2817657 is committed.
-   */
-  protected function assertOrderInPage(array $items) {
-    $session = $this->getSession();
-    $text = $session->getPage()->getHtml();
-    $strings = [];
-    foreach ($items as $item) {
-      if (($pos = strpos($text, $item)) === FALSE) {
-        throw new ExpectationException("Cannot find '$item' in the page", $session->getDriver());
-      }
-      $strings[$pos] = $item;
-    }
-    ksort($strings);
-    $ordered = implode(', ', array_map(function ($item) {
-      return "'$item'";
-    }, $items));
-    $this->assertSame($items, array_values($strings), "Found strings, ordered as: $ordered.");
-  }
-
   /**
    * Tests book outline AJAX request.
    */
diff --git a/core/modules/layout_builder/tests/src/FunctionalJavascript/ContentPreviewToggleTest.php b/core/modules/layout_builder/tests/src/FunctionalJavascript/ContentPreviewToggleTest.php
index 70597c615d..f154cc0405 100644
--- a/core/modules/layout_builder/tests/src/FunctionalJavascript/ContentPreviewToggleTest.php
+++ b/core/modules/layout_builder/tests/src/FunctionalJavascript/ContentPreviewToggleTest.php
@@ -94,7 +94,7 @@ public function testContentPreviewToggle() {
     $this->assertContextualLinks();
 
     // Confirm repositioning blocks works with content preview disabled.
-    $this->assertOrderInPage([$links_field_placeholder_label, $body_field_placeholder_label]);
+    $this->assertSession()->responseContentHasOrder([$links_field_placeholder_label, $body_field_placeholder_label]);
 
     $region_content = '.layout__region--content';
     $links_block = "[data-layout-content-preview-placeholder-label='$links_field_placeholder_label']";
@@ -110,7 +110,7 @@ public function testContentPreviewToggle() {
     $assert_session->pageTextNotContains($content_preview_body_text);
 
     // Check that drag successfully repositioned blocks.
-    $this->assertOrderInPage([$body_field_placeholder_label, $links_field_placeholder_label]);
+    $this->assertSession()->responseContentHasOrder([$body_field_placeholder_label, $links_field_placeholder_label]);
 
     // Check if block position maintained after enabling content preview.
     $this->assertTrue($page->hasUncheckedField('layout-builder-content-preview'));
@@ -118,7 +118,7 @@ public function testContentPreviewToggle() {
     $this->assertNotEmpty($assert_session->waitForText($content_preview_body_text));
     $assert_session->pageTextContains($content_preview_body_text);
     $this->assertNotEmpty($assert_session->waitForText('Placeholder for the "Links" field'));
-    $this->assertOrderInPage([$content_preview_body_text, 'Placeholder for the "Links" field']);
+    $this->assertSession()->responseContentHasOrder([$content_preview_body_text, 'Placeholder for the "Links" field']);
   }
 
   /**
@@ -136,24 +136,4 @@ protected function assertContextualLinks() {
     $assert_session->assertNoElementAfterWait('css', '#drupal-off-canvas');
   }
 
-  /**
-   * Asserts that blocks in a given order in the page.
-   *
-   * @param string[] $items
-   *   An ordered list of strings that should appear in the blocks.
-   */
-  protected function assertOrderInPage(array $items) {
-    $session = $this->getSession();
-    $page = $session->getPage();
-    $blocks = $page->findAll('css', '[data-layout-content-preview-placeholder-label]');
-
-    // Filter will only return value if block contains expected text.
-    $blocks_with_expected_text = array_filter($blocks, function ($block, $key) use ($items) {
-      $block_text = $block->getText();
-      return strpos($block_text, $items[$key]) !== FALSE;
-    }, ARRAY_FILTER_USE_BOTH);
-
-    $this->assertCount(count($items), $blocks_with_expected_text);
-  }
-
 }
diff --git a/core/modules/system/tests/modules/test_page_test/src/Controller/Test.php b/core/modules/system/tests/modules/test_page_test/src/Controller/Test.php
index 19e0bffe29..49fc31884a 100644
--- a/core/modules/system/tests/modules/test_page_test/src/Controller/Test.php
+++ b/core/modules/system/tests/modules/test_page_test/src/Controller/Test.php
@@ -233,4 +233,15 @@ public function deprecations() {
     ];
   }
 
+  /**
+   * Renders a page with strings in a specific order.
+   *
+   * @return array
+   *   A render array as expected by
+   *   \Drupal\Core\Render\RendererInterface::render().
+   */
+  public function renderPageWithOrderedContent() {
+    return ['#markup' => 'item3<p class="item1 item2 item3">item2</p>item1'];
+  }
+
 }
diff --git a/core/modules/system/tests/modules/test_page_test/test_page_test.routing.yml b/core/modules/system/tests/modules/test_page_test/test_page_test.routing.yml
index f3744df5ee..368f8fa954 100644
--- a/core/modules/system/tests/modules/test_page_test/test_page_test.routing.yml
+++ b/core/modules/system/tests/modules/test_page_test/test_page_test.routing.yml
@@ -152,3 +152,11 @@ test_page_test.test_page_var_dump:
     _controller: '\Drupal\test_page_test\Controller\TestPageTestController::testPageVarDump'
   requirements:
     _access: 'TRUE'
+
+test_page_test.order_in_page:
+  path: '/test-order-in-page'
+  defaults:
+    _title: 'Page with strings in a given order'
+    _controller: '\Drupal\test_page_test\Controller\Test::renderPageWithOrderedContent'
+  requirements:
+    _access: 'TRUE'
diff --git a/core/tests/Drupal/FunctionalJavascriptTests/TableDrag/TableDragTest.php b/core/tests/Drupal/FunctionalJavascriptTests/TableDrag/TableDragTest.php
index cf36dc1929..d01b8cc965 100644
--- a/core/tests/Drupal/FunctionalJavascriptTests/TableDrag/TableDragTest.php
+++ b/core/tests/Drupal/FunctionalJavascriptTests/TableDrag/TableDragTest.php
@@ -3,7 +3,6 @@
 namespace Drupal\FunctionalJavascriptTests\TableDrag;
 
 use Behat\Mink\Element\NodeElement;
-use Behat\Mink\Exception\ExpectationException;
 use Drupal\FunctionalJavascriptTests\WebDriverTestBase;
 
 /**
@@ -110,7 +109,7 @@ public function testDragAndDrop() {
     $weight_select3 = $page->findField("table[3][weight]");
 
     // Check that initially the rows are in the correct order.
-    $this->assertOrder(['Row with id 1', 'Row with id 2', 'Row with id 3']);
+    $this->assertSession()->pageTextHasOrder(['Row with id 1', 'Row with id 2', 'Row with id 3']);
 
     // Check that the 'unsaved changes' text is not present in the message area.
     $this->assertSession()->pageTextNotContains('You have unsaved changes.');
@@ -126,7 +125,7 @@ public function testDragAndDrop() {
     $this->assertSession()->waitForText('You have unsaved changes.');
 
     // Check that row1 and row2 were swapped.
-    $this->assertOrder(['Row with id 2', 'Row with id 1', 'Row with id 3']);
+    $this->assertSession()->pageTextHasOrder(['Row with id 2', 'Row with id 1', 'Row with id 3']);
 
     // Check that weights were changed.
     $this->assertGreaterThan($weight_select2->getValue(), $weight_select1->getValue());
@@ -137,7 +136,7 @@ public function testDragAndDrop() {
     $row3->dragTo($row1);
 
     // Check that the order is: row2, row3 and row1.
-    $this->assertOrder(['Row with id 2', 'Row with id 3', 'Row with id 1']);
+    $this->assertSession()->pageTextHasOrder(['Row with id 2', 'Row with id 3', 'Row with id 1']);
   }
 
   /**
@@ -311,31 +310,6 @@ public function testTableDragChangedWarning() {
     $this->assertSession()->pageTextContainsOnce('You have unsaved changes.');
   }
 
-  /**
-   * Asserts that several pieces of markup are in a given order in the page.
-   *
-   * @param string[] $items
-   *   An ordered list of strings.
-   *
-   * @throws \Behat\Mink\Exception\ExpectationException
-   *   When any of the given string is not found.
-   *
-   * @todo Remove this and use the WebAssert method when #2817657 is done.
-   */
-  protected function assertOrder(array $items) {
-    $session = $this->getSession();
-    $text = $session->getPage()->getHtml();
-    $strings = [];
-    foreach ($items as $item) {
-      if (($pos = strpos($text, $item)) === FALSE) {
-        throw new ExpectationException("Cannot find '$item' in the page", $session->getDriver());
-      }
-      $strings[$pos] = $item;
-    }
-    ksort($strings);
-    $this->assertSame($items, array_values($strings), "Strings found on the page but incorrectly ordered.");
-  }
-
   /**
    * Tests nested draggable tables through keyboard.
    */
diff --git a/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php b/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php
index c34c61f106..12909ad6cb 100644
--- a/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php
+++ b/core/tests/Drupal/FunctionalTests/BrowserTestBaseTest.php
@@ -997,4 +997,55 @@ public function testVarDump() {
     $this->assertStringContainsString('</samp>}', $body);
   }
 
+  /**
+   * Tests assertion of a substring sequence in a string, page content or text.
+   *
+   * @covers \Drupal\Tests\WebAssert::orderInString
+   * @covers \Drupal\Tests\WebAssert::responseContentHasOrder
+   * @covers \Drupal\Tests\WebAssert::pageTextHasOrder
+   */
+  public function testSubstringsOrder() {
+    // Check that asserting order in an arbitrary string works.
+    $this->assertSession()->orderInString(['item3', 'item1', 'item2'], '(*&^^ item3.,.><> item1@...... item2');
+
+    // Check again with a repeating substring.
+    $this->assertSession()->orderInString([
+      'item1',
+      'item2',
+      'item1',
+      'item1',
+      'item2',
+    ], 'item1.item2.item1.item1.item2');
+
+    // Check that passing existing substrings in the wrong order fails.
+    try {
+      $this->assertSession()->orderInString(['item2', 'item1', 'item3'], 'item1item3item2');
+      $this->fail('Expected ExpectationFailedException exception has not been thrown.');
+    }
+    catch (ExpectationFailedException $exception) {
+      // Expected exception; just continue testing.
+    }
+
+    // Check that passing empty array of substrings works.
+    $this->assertSession()->orderInString([], 'item1item3item2');
+
+    // Check that passing array with one substring works.
+    $this->assertSession()->orderInString(['item1'], 'item1item3item2');
+
+    $this->drupalGet('test-order-in-page');
+
+    // Check that asserting order in page markup and page text works.
+    $this->assertSession()->responseContentHasOrder(['item3', 'item1', 'item2']);
+    $this->assertSession()->pageTextHasOrder(['item3', 'item2', 'item1']);
+
+    // Check that passing non-existent substrings throws an exception.
+    try {
+      $this->assertSession()->responseContentHasOrder(['item5', 'item1', 'item8']);
+      $this->fail('Expected ElementNotFoundException exception has not been thrown.');
+    }
+    catch (ElementNotFoundException $exception) {
+      $this->assertSame("Substring(s): 'item5', 'item8' not found.", $exception->getMessage());
+    }
+  }
+
 }
diff --git a/core/tests/Drupal/Tests/WebAssert.php b/core/tests/Drupal/Tests/WebAssert.php
index 1856ab5ffb..a52c025719 100644
--- a/core/tests/Drupal/Tests/WebAssert.php
+++ b/core/tests/Drupal/Tests/WebAssert.php
@@ -1096,4 +1096,80 @@ public function checkboxNotChecked($field, TraversableElement $container = NULL)
     return parent::checkboxNotChecked($field, $container);
   }
 
+  /**
+   * Asserts that several strings are in a given order inside a string.
+   *
+   * @param string[] $expected_order
+   *   An ordered list of strings.
+   * @param string $string
+   *   The string to be searched.
+   *
+   * @throws \Behat\Mink\Exception\ElementNotFoundException
+   *   When one or more strings from the list are not found in $string.
+   * @throws \PHPUnit\Framework\ExpectationFailedException
+   *   When the strings from list are ordered incorrectly in $string.
+   */
+  public function orderInString(array $expected_order, $string) {
+    $not_found = [];
+    foreach ($expected_order as $substring) {
+      if (strpos($string, $substring) === FALSE) {
+        $not_found[] = "'$substring'";
+      }
+    }
+    if ($not_found) {
+      throw new ElementNotFoundException($this->session->getDriver(), sprintf('Substring(s): %s', implode(', ', array_unique($not_found))));
+    }
+
+    $actual_order = [];
+    $offset = 0;
+    foreach ($expected_order as $substring) {
+      if (($pos = strpos($string, $substring, $offset)) !== FALSE) {
+        $actual_order[$pos] = $substring;
+        $offset = $pos + strlen($substring);
+      }
+    }
+    ksort($actual_order);
+
+    Assert::assertSame(array_values($expected_order), array_values($actual_order), 'Strings found but incorrectly ordered');
+  }
+
+  /**
+   * Asserts that several strings are in a given order in the response content.
+   *
+   * @param string[] $expected_order
+   *   An ordered list of strings.
+   *
+   * @throws \Behat\Mink\Exception\ElementNotFoundException
+   *   When one or more strings from the list are not found in page HTML.
+   * @throws \Behat\Mink\Exception\ExpectationException
+   *   When the strings from list are ordered incorrectly in page HTML.
+   */
+  public function responseContentHasOrder(array $expected_order) {
+    $this->orderInString($expected_order, $this->session->getPage()->getContent());
+  }
+
+  /**
+   * Asserts that multiple strings are in a given order in the page text.
+   *
+   * @param string[] $expected_order
+   *   An ordered list of strings.
+   *
+   * @throws \Behat\Mink\Exception\ElementNotFoundException
+   *   When one or more strings from the list are not found in page text.
+   * @throws \Behat\Mink\Exception\ExpectationException
+   *   When the strings from list are ordered incorrectly in page text.
+   */
+  public function pageTextHasOrder(array $expected_order) {
+    try {
+      $this->orderInString($expected_order, $this->session->getPage()
+        ->getText());
+    }
+    catch (ExpectationException $exception) {
+      // When checking the order in page text we throw ResponseTextException
+      // rather than ExpectationException, since ResponseTextException is more
+      // specialized and fitting.
+      throw new ResponseTextException($exception->getMessage(), $this->session->getDriver(), $exception);
+    }
+  }
+
 }
