diff --git a/core/modules/media_library/src/Form/FileUploadForm.php b/core/modules/media_library/src/Form/FileUploadForm.php
index 2c6dedff82..1bfcf5b215 100644
--- a/core/modules/media_library/src/Form/FileUploadForm.php
+++ b/core/modules/media_library/src/Form/FileUploadForm.php
@@ -158,6 +158,16 @@ protected function buildInputElement(array $form, FormStateInterface $form_state
       '#remaining_slots' => $slots,
     ];
 
+    // Add upload resolution validation if file is an image.
+    $fieldSettings = $item->getFieldDefinition()->getSettings();
+    $form['container']['upload']['#upload_validators']['file_validate_is_image'] = [];
+    if (isset($fieldSettings['max_resolution']) || isset($fieldSettings['min_resolution'])) {
+      $form['container']['upload']['#upload_validators']['file_validate_image_resolution'] = [
+        $fieldSettings['max_resolution'] ?? 0,
+        $fieldSettings['min_resolution'] ?? 0,
+      ];
+    }
+
     $file_upload_help = [
       '#theme' => 'file_upload_help',
       '#upload_validators' => $form['container']['upload']['#upload_validators'],
